File: textbook/python/01-dataflow/04_sync_express.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — DataFlow / SyncExpress CRUD
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Use SyncExpress for synchronous CRUD without async/await
# LEVEL: Intermediate
# PARITY: Full — both languages have sync wrappers
# VALIDATES: SyncExpress via db.express_sync, same CRUD operations as
#            async but synchronous (create, read, update, delete, list, count)
#
# Run: uv run python textbook/python/01-dataflow/04_sync_express.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import asyncio
import os
import tempfile

from dataflow import DataFlow

# ── Setup ───────────────────────────────────────────────────────────
# SyncExpress wraps the async ExpressDataFlow methods with synchronous
# equivalents. Useful in CLI scripts, synchronous FastAPI handlers,
# pytest without asyncio, and other non-async code.
#
# Internally, SyncExpress maintains a persistent event loop in a
# background daemon thread so that database connections survive
# across multiple sync calls.

db_path = os.path.join(tempfile.gettempdir(), "textbook_dataflow_04.db")
db = DataFlow(database_url=f"sqlite:///{db_path}")


@db.model
class Book:
    title: str
    author: str
    pages: int
    available: bool = True


# initialize() is still async — call it once before using express_sync
async def setup():
    await db.initialize()


asyncio.run(setup())

# ── 1. ACCESS — db.express_sync property ────────────────────────────
# After initialize(), access SyncExpress through db.express_sync.
# No 'await' needed — all calls are plain synchronous functions.

sync = db.express_sync
assert sync is not None, "express_sync property returns SyncExpress instance"

# ── 2. CREATE — insert records synchronously ────────────────────────

book1 = sync.create(
    "Book",
    {
        "title": "The Pragmatic Programmer",
        "author": "Hunt & Thomas",
        "pages": 352,
    },
)

assert "id" in book1, "Created record gets an auto-generated id"
assert book1["title"] == "The Pragmatic Programmer"
assert book1["available"] is True, "Default value applied"
print(f"Created: {book1['title']}")

book2 = sync.create(
    "Book",
    {
        "title": "Clean Code",
        "author": "Robert C. Martin",
        "pages": 464,
    },
)
book3 = sync.create(
    "Book",
    {
        "title": "Refactoring",
        "author": "Martin Fowler",
        "pages": 448,
        "available": False,
    },
)

# ── 3. READ — fetch a single record by id ──────────────────────────

fetched = sync.read("Book", book1["id"])

assert fetched is not None, "Record found by id"
assert fetched["title"] == "The Pragmatic Programmer"
print(f"Read: {fetched['title']}")

# Read non-existent record returns None
missing = sync.read("Book", "nonexistent-id-999")
assert missing is None, "Missing record returns None"

# ── 4. UPDATE — modify an existing record ───────────────────────────

updated = sync.update("Book", book3["id"], {"available": True})

assert updated["available"] is True, "Field updated"
print(f"Updated: {updated['title']} is now available")

# Verify persistence
refetched = sync.read("Book", book3["id"])
assert refetched["available"] is True, "Update persisted"

# ── 5. DELETE — remove a record ─────────────────────────────────────

deleted = sync.delete("Book", book2["id"])
assert deleted is True, "Delete returns True on success"

gone = sync.read("Book", book2["id"])
assert gone is None, "Deleted record returns None on read"

# ── 6. LIST — query multiple records ────────────────────────────────

all_books = sync.list("Book")
assert len(all_books) == 2, "Two books remain after delete"

# List with ordering
ordered = sync.list("Book", order_by="pages")
assert len(ordered) == 2
pages = [b["pages"] for b in ordered]
assert pages == sorted(pages), "Ordered by pages ascending"

# ── 7. COUNT — count records ────────────────────────────────────────

total = sync.count("Book")
assert total == 2, "Two records in total"

# ── 8. Same instance reuse ──────────────────────────────────────────
# The SyncExpress instance is cached on the DataFlow object.
# Accessing db.express_sync again returns the same instance.

sync2 = db.express_sync
assert sync2 is sync, "Same SyncExpress instance returned"

# ── Cleanup ─────────────────────────────────────────────────────────


async def teardown():
    await db.stop()


asyncio.run(teardown())

try:
    os.unlink(db_path)
except OSError:
    pass

print("PASS: 01-dataflow/04_sync_express")

File: textbook/rust/01-dataflow/src/bin/04_sync_express.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- DataFlow / Sync Express
//!
//! OBJECTIVE: Use DataFlowExpressSync for blocking CRUD in non-async contexts.
//! LEVEL: Basic
//! PARITY: Equivalent -- Python has synchronous wrappers via run_sync();
//!         Rust provides DataFlowExpressSync with identical API surface.
//! VALIDATES: DataFlowExpressSync, create, read, update, delete, list
//!
//! Run: cargo run -p tutorial-dataflow --bin 04_sync_express

use kailash_dataflow::prelude::*;
use kailash_value::{Value, ValueMap};
use std::sync::Arc;

// DataFlowExpressSync wraps the async express API with blocking calls.
// Useful for CLI tools, scripts, and contexts where async is not available.

#[tokio::main]
async fn main() {
    // ── Setup ──
    let mut df = DataFlow::new("sqlite::memory:")
        .await
        .expect("connect");

    let model = ModelDefinition::new("Note", "notes")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("title", FieldType::Text, |f| f.required())
        .field("body", FieldType::Text, |f| f)
        .auto_timestamps();

    df.register_model(model).expect("register");

    // ── Get Sync Express Handle ──
    // DataFlowExpressSync provides the same API as DataFlowExpress
    // but blocks the calling thread instead of returning futures.

    let sync_express = df.express_sync();

    // ── Create (Sync) ──
    let mut data = ValueMap::new();
    data.insert(Arc::from("title"), Value::from("Meeting notes"));
    data.insert(Arc::from("body"), Value::from("Discussed roadmap"));

    let created = sync_express.create("Note", data).expect("sync create");
    let note_id = created
        .get("id" as &str)
        .expect("id")
        .as_i64()
        .expect("integer id");
    assert!(note_id > 0);

    // ── Read (Sync) ──
    let note = sync_express
        .read("Note", &note_id.to_string())
        .expect("sync read");
    assert_eq!(note.get("title" as &str), Some(&Value::from("Meeting notes")));

    // ── Update (Sync) ──
    let mut updates = ValueMap::new();
    updates.insert(Arc::from("body"), Value::from("Discussed roadmap and milestones"));

    let updated = sync_express
        .update("Note", &note_id.to_string(), updates)
        .expect("sync update");
    assert_eq!(
        updated.get("body" as &str),
        Some(&Value::from("Discussed roadmap and milestones"))
    );

    // ── List (Sync) ──
    // Create a few more notes first.
    for title in ["Design doc", "Sprint review", "Retrospective"] {
        let mut data = ValueMap::new();
        data.insert(Arc::from("title"), Value::from(title));
        sync_express.create("Note", data).expect("create");
    }

    let all = sync_express
        .list("Note", ListOptions::default())
        .expect("sync list");
    assert_eq!(all.len(), 4);

    // ── Delete (Sync) ──
    sync_express
        .delete("Note", &note_id.to_string())
        .expect("sync delete");

    let remaining = sync_express
        .list("Note", ListOptions::default())
        .expect("sync list after delete");
    assert_eq!(remaining.len(), 3);

    println!("PASS: 01-dataflow/04_sync_express");
}
